internal/handlers/tigris/msg_delete.go

```diff
@@ -47,55 +47,61 @@ func (h *Handler) MsgDelete(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
 		return nil, err
 	}
 
+	ordered := true
+	if ordered, err = common.GetOptionalParam(document, "ordered", ordered); err != nil {
+		return nil, err
+	}
+
 	var deleted int32
-	for i := 0; i < deletes.Len(); i++ {
+	processQuery := func(i int) error {
 		d, err := common.AssertType[*types.Document](must.NotFail(deletes.Get(i)))
 		if err != nil {
-			return nil, err
+			return err
 		}
 
 		if err := common.Unimplemented(d, "collation", "hint"); err != nil {
-			return nil, err
+			return err
 		}
 
 		var filter *types.Document
 		if filter, err = common.GetOptionalParam(d, "q", filter); err != nil {
-			return nil, err
+			return err
 		}
 
+		// TODO https://github.com/FerretDB/FerretDB/issues/982
 		var limit int64
 		if l, _ := d.Get("limit"); l != nil {
 			if limit, err = common.GetWholeNumberParam(l); err != nil {
-				return nil, err
+				return err
 			}
 		}
 
 		var fp fetchParam
 		if fp.db, err = common.GetRequiredParam[string](document, "$db"); err != nil {
-			return nil, err
+			return err
 		}
 		collectionParam, err := document.Get(document.Command())
 		if err != nil {
-			return nil, err
+			return err
 		}
 		var ok bool
 		if fp.collection, ok = collectionParam.(string); !ok {
-			return nil, common.NewErrorMsg(
+			return common.NewErrorMsg(
 				common.ErrBadValue,
 				fmt.Sprintf("collection name has invalid type %s", common.AliasFromType(collectionParam)),
 			)
 		}
 
 		fetchedDocs, err := h.fetch(ctx, fp)
```

Comment:
It looks like with the current approach, you fetch all the documents for each iteration. I don't think it's reasonable to do it. Let's not fetch all the documents more than once per single query.

Reply:
I didn't change any logic in the code inside the loop. All I've done was just extracting it to the nested function. However, as I'm looking at the code I think that it's reasonable to fetch documents for each iteration.

## Example

Let's say for an example that we have a collection with elements: 
```
[
  { _id: "a", v: "foo"}, 
  { _id: "b", v: "foo"}, 
  { _id: "c", v: "bar"}
]
```
We want to run a delete query with  the following filters: 1: `{"_id":"a"}` 2: `{"v":"foo"}`

If we implement document fetching outside the iteration it will execute in that way:
### fetch documents:
**docs**: `[{ _id: "a", v: "foo"}, { _id: "b", v: "foo"}, { _id: "c", v: "bar"}]`

### first iteration
**docs**: `[{ _id: "a", v: "foo"}, { _id: "b", v: "foo"}, { _id: "c", v: "bar"}]`
**query**: `{"_id":"a"}`
**delete will remove field:**: `{ _id: "a", v: "foo"}`

### second iteration
**docs**: `[{ _id: "a", v: "foo"}, { _id: "b", v: "foo"}, { _id: "c", v: "bar"}]`
**query**: `{"_id":"a"}`
**delete will **try to** remove fields:**: `{ _id: "a", v: "foo"}, { _id: "b", v: "foo"}`, but the first one will just try to remove the query that was already removed in the **first iteration**.

Fetching the document on every iteration solves the problem.

Reply:
Yeah, I think you are right.